lib/manifest/manifest.go

```diff
@@ -0,0 +1,93 @@
+// Package manifest collects functions about manifests that are going to be lifted to library-go
+// https://github.com/openshift/library-go
+// Thus it should not depend on any packages from CVO: github.com/openshift/cluster-version-operator/
+package manifest
+
+import (
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/library-go/pkg/manifest"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+)
+
+// InclusionConfiguration configures manifest inclusion, so
+// callers can opt in to new filtering options instead of having to
+// update existing call-sites, even if they do not need a new
+// filtering option.
+type InclusionConfiguration struct {
+	// ExcludeIdentifier, if non-nil, excludes manifests that match the exclusion identifier.
+	ExcludeIdentifier *string
+
+	// RequiredFeatureSet, if non-nil, excludes manifests unless they match the desired feature set.
+	RequiredFeatureSet *string
+
+	// Profile, if non-nil, excludes manifests unless they match the cluster profile.
+	Profile *string
+
+	// Capabilities, if non-nil, excludes manifests unless they match the enabled cluster capabilities.
+	Capabilities *configv1.ClusterVersionCapabilitiesStatus
+
+	// Overrides excludes manifests for overridden resources.
+	Overrides []configv1.ComponentOverride
+
+	// Platform, if non-nil, excludes CredentialsRequests manifests unless they match the infrastructure platform.
+	Platform *string
+}
+
+// GetImplicitlyEnabledCapabilities returns a set of capabilities that are implicitly enabled after a cluster update.
+// The arguments are two sets of manifests, manifest inclusion configuration, and
+// a set of capabilities that are implicitly enabled on the cluster, i.e., the capabilities
+// that are NOT specified in the cluster version but has to considered enabled on the cluster.
+// The manifest inclusion configuration is used to determine if a manifest should be included.
+// In other words, whether, or not the cluster version operator reconcile that manifest on the cluster.
+// The two sets of manifests are respectively from the release that is currently running on the cluster and
+// from the release that the cluster is updated to.
+func GetImplicitlyEnabledCapabilities(
```

Comment:
How do we test this implementation is:

1. Correct?
2. Equivalent to the deleted `GetImplicitlyEnabledCapabilities`?

For (1) I assume that we'll need tests once we lift it to o/library-go so why not write them right away?
For (2) I wonder if if would be proper to keep the old implementation for a while, implement a test that shows that the two implementations are equivalent (same inputs -> same outputs) and only delete the code (and the equivalence test) once the o/library-go implementation fully settles?

Reply:
For (2): There are tests for it already.

https://github.com/openshift/cluster-version-operator/blob/198fdabfbbc09bf03213c18f0993e759de77ca26/pkg/payload/payload_test.go#L220

At least, from those tests point of view, they are equivalent.

For (1):

I moved over the unit tests from (2) to https://github.com/openshift/library-go/pull/1908/files with modifications
- to fit the new function signature (added a new arg `manifestInclusionConfiguration`), and
- to avoid copying [testdata/payloadcapabilitytest](https://github.com/openshift/cluster-version-operator/tree/5fd85809c592eb66a099804512b05f757b2bb2fd/pkg/cvo/testdata/payloadcapabilitytest) there and convincing `o/lib` approvers that it OK to do it .

Reply:
Ah okay, we have tests on the payload-level `GetImplicitlyEnabledCapabilities` which calls the new function, good!

:shipit: then